Unable to solve NameExpr in Annotation

I have the following class
```java
@interface Foo { 
    int a = 0; 
    int b = a; 
}
```
When i try to resolve the reference, it throws a UnsupportedOperationException. This happens because method `getAllFields` is not implemented `JavaParserAnnotationDeclaration` and according to issue https://github.com/javaparser/javaparser/issues/1837 annotations should not have fields. The problem gets confusing when the AST parses the annotation with a field declaration on it.

Should it be consider a problem with  `JavaParserAnnotationDeclaration` or with the parser?

#### AST from the example
![graphviz](https://user-images.githubusercontent.com/5434488/108416838-28e02000-7227-11eb-88c6-2f34b6f8147f.png)

#### Stack trace
```
Caused by: java.lang.UnsupportedOperationException
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserAnnotationDeclaration.getAllFields(JavaParserAnnotationDeclaration.java:68)
	at com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration.hasField(ResolvedReferenceTypeDeclaration.java:150)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AnnotationDeclarationContext.solveSymbol(AnnotationDeclarationContext.java:53)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.solveSymbolInParentContext(AbstractJavaParserContext.java:184)
	at com.github.javaparser.symbolsolver.core.resolution.Context.solveSymbol(Context.java:115)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbol(SymbolSolver.java:70)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbol(SymbolSolver.java:74)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:171)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:181)
	at com.github.javaparser.ast.expr.NameExpr.resolve(NameExpr.java:170)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:72)
```

#### Unit test
```java
@Test
void test() {
  StaticJavaParser.getConfiguration()
		.setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver()));

  String sourceCode = "@interface Foo { int a = 0; int b = a; }";

  CompilationUnit cu = StaticJavaParser.parse(sourceCode);

  Optional<NameExpr> nameExpr = cu.findFirst(NameExpr.class);
  assertTrue(nameExpr.isPresent());
  assertDoesNotThrow(nameExpr.get()::resolve);
}
```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnnotationDeclaration.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnnotationDeclaration.java
@@ -30,11 +30,7 @@ import com.github.javaparser.symbolsolver.logic.AbstractTypeDeclaration;
 import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
 import com.github.javaparser.symbolsolver.model.typesystem.ReferenceTypeImpl;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
 import java.util.stream.Collectors;
 
 /**
@@ -59,8 +55,10 @@ public class JavaParserAnnotationDeclaration extends AbstractTypeDeclaration imp
 
     @Override
     public List<ResolvedFieldDeclaration> getAllFields() {
-        // TODO #1837
-        throw new UnsupportedOperationException();
+         return wrappedNode.getFields().stream()
+                .flatMap(field -> field.getVariables().stream())
+                .map(var -> new JavaParserFieldDeclaration(var, typeSolver))
+                .collect(Collectors.toList());
     }
 
     @Override

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javassistmodel/JavassistAnnotationDeclaration.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javassistmodel/JavassistAnnotationDeclaration.java
@@ -90,8 +90,7 @@ public class JavassistAnnotationDeclaration extends AbstractTypeDeclaration impl
 
     @Override
     public List<ResolvedFieldDeclaration> getAllFields() {
-        // TODO #1837
-        throw new UnsupportedOperationException();
+        return javassistTypeDeclarationAdapter.getDeclaredFields();
     }
 
     @Override

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclaration.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclaration.java
@@ -135,8 +135,7 @@ public class ReflectionAnnotationDeclaration extends AbstractTypeDeclaration imp
 
     @Override
     public List<ResolvedFieldDeclaration> getAllFields() {
-        // TODO #1837
-        throw new UnsupportedOperationException();
+        return reflectionClassAdapter.getAllFields();
     }
 
     @Override

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnnotationDeclarationTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnnotationDeclarationTest.java
@@ -1,0 +1,90 @@
+/*
+ * Copyright (C) 2007-2010 Júlio Vilmar Gesser.
+ * Copyright (C) 2011, 2013-2021 The JavaParser Team.
+ *
+ * This file is part of JavaParser.
+ *
+ * JavaParser can be used either under the terms of
+ * a) the GNU Lesser General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ * b) the terms of the Apache License
+ *
+ * You should have received a copy of both licenses in LICENCE.LGPL and
+ * LICENCE.APACHE. Please refer to those files for details.
+ *
+ * JavaParser is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+package com.github.javaparser.symbolsolver.javaparsermodel.declarations;
+
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ParseResult;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.body.AnnotationDeclaration;
+import com.github.javaparser.ast.expr.NameExpr;
+import com.github.javaparser.resolution.declarations.ResolvedFieldDeclaration;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class JavaParserAnnotationDeclarationTest extends AbstractResolutionTest {
+
+	private final TypeSolver typeSolver = new ReflectionTypeSolver();
+	private final JavaParser javaParser = createParserWithResolver(typeSolver);
+
+	@Test
+	void getAllFields_shouldReturnASingleField() {
+		String sourceCode = "@interface Foo { int a = 0; }";
+
+		ParseResult<CompilationUnit> result = javaParser.parse(sourceCode);
+		assertTrue(result.getResult().isPresent());
+		CompilationUnit cu = result.getResult().get();
+
+		Optional<AnnotationDeclaration> annotation = cu.findFirst(AnnotationDeclaration.class);
+		assertTrue(annotation.isPresent());
+
+		List<ResolvedFieldDeclaration> fields = annotation.get().resolve().getAllFields();
+		assertEquals(1, fields.size());
+		assertEquals("a", fields.get(0).getName());
+	}
+
+	@Test
+	void getAllFields_shouldReturnMultipleVariablesDeclaration() {
+		String sourceCode = "@interface Foo { int a = 0, b = 1; }";
+
+		ParseResult<CompilationUnit> result = javaParser.parse(sourceCode);
+		assertTrue(result.getResult().isPresent());
+		CompilationUnit cu = result.getResult().get();
+
+		Optional<AnnotationDeclaration> annotation = cu.findFirst(AnnotationDeclaration.class);
+		assertTrue(annotation.isPresent());
+
+		List<ResolvedFieldDeclaration> fields = annotation.get().resolve().getAllFields();
+		assertEquals(2, fields.size());
+		assertEquals("a", fields.get(0).getName());
+		assertEquals("b", fields.get(1).getName());
+	}
+
+	@Test
+	void testForIssue3094() {
+		String sourceCode = "@interface Foo { int a = 0; int b = a; }";
+		ParseResult<CompilationUnit> result = javaParser.parse(sourceCode);
+		assertTrue(result.getResult().isPresent());
+		CompilationUnit cu = result.getResult().get();
+
+		Optional<NameExpr> nameExpr = cu.findFirst(NameExpr.class);
+		assertTrue(nameExpr.isPresent());
+		assertDoesNotThrow(nameExpr.get()::resolve);
+	}
+
+}

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/javassistmodel/JavassistAnnotationDeclarationTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/javassistmodel/JavassistAnnotationDeclarationTest.java
@@ -57,13 +57,6 @@ class JavassistAnnotationDeclarationTest extends AbstractTypeDeclarationTest imp
         super.containerTypeCantBeNull();
     }
 
-    @Disabled(value = "This feature is not yet implemented. See https://github.com/javaparser/javaparser/issues/1837")
-    @Test
-    @Override
-    public void getAllFieldsCantBeNull() {
-        super.getAllFieldsCantBeNull();
-    }
-
     @Disabled(value = "This feature is not yet implemented. See https://github.com/javaparser/javaparser/issues/1838")
     @Test
     @Override

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclarationTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclarationTest.java
@@ -21,17 +21,18 @@
 
 package com.github.javaparser.symbolsolver.reflectionmodel;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-
 import com.github.javaparser.resolution.declarations.ResolvedDeclaration;
 import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
 import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
 import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+import org.junit.jupiter.api.Test;
+
 import java.util.Collections;
 import java.util.stream.Collectors;
-import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 
 @interface OuterAnnotation {
   @interface InnerAnnotation {}
@@ -39,6 +40,10 @@ import org.junit.jupiter.api.Test;
 
 @interface WithValue {
   String value();
+}
+
+@interface WithField {
+  int FIELD_DECLARATION = 0;
 }
 
 class ReflectionAnnotationDeclarationTest {
@@ -77,4 +82,14 @@ class ReflectionAnnotationDeclarationTest {
         annotation.solveMethod("value", Collections.emptyList(), false);
     assertEquals("value", symbolReference.getCorrespondingDeclaration().getName());
   }
+
+  @Test
+  void getAllFields_shouldReturnTheCorrectFields() {
+    ReflectionAnnotationDeclaration annotation =
+            (ReflectionAnnotationDeclaration) typeSolver.solveType(
+                    "com.github.javaparser.symbolsolver.reflectionmodel.WithField");
+    assertEquals(Collections.singleton("FIELD_DECLARATION"),
+            annotation.getAllFields().stream().map(ResolvedDeclaration::getName).collect(Collectors.toSet()));
+  }
+
 }

